### PR TITLE
Reify AST node types and remove unneeded nullability

### PR DESCRIPTION
**Pull request by spawnia**
This achieves a few things:

- make the code easier to understand both by humans and static analyzers
- improve performance by eliminating unnecessary null checks
- improve usability through a clarified API

### CODE REVIEW

File: src/Utils/SchemaExtender.php [hunk 1]
```diff
@@ -615,13 +616,13 @@ static function (string $typeName) use ($schema) {
         $types = array_merge(
             // Iterate through all types, getting the type definition for each, ensuring
             // that any type not directly referenced by a field will get created.
-            array_map(static function ($type) {
+            array_map(static function (Type $type) : Type {
                 return static::extendNamedType($type);
-            }, array_values($schema->getTypeMap())),
+            }, $schema->getTypeMap()),
```
Comment (simPod) [inline src/Utils/SchemaExtender.php:621]: didn't resetting of those array indexes through `array_values()` play some part?

File: src/Utils/ASTDefinitionBuilder.php [hunk 2]
```diff
@@ -318,95 +319,89 @@ public function buildField(FieldDefinitionNode $field)
...
-                ? function () use ($def) {
-                    return Utils::map(
-                        $def->types,
-                        function ($typeNode) : Type {
-                            return $this->buildType($typeNode);
-                        }
-                    );
-                }
-                : [],
+            'types'       => function () use ($def) {
```
Comment (simPod) [inline src/Utils/ASTDefinitionBuilder.php:392]: we can add return type hint here as well

File: src/Utils/ASTDefinitionBuilder.php [hunk 3]
```diff
@@ -41,28 +43,28 @@
...
     private $resolveType;
 
-    /** @var Type[] */
+    /** @var array<string, Type> */
     private $cache;
 
     /**
-     * @param Node[] $typeDefinitionsMap
-     * @param bool[] $options
+     * @param array<string, Node  &TypeDefinitionNode> $typeDefinitionsMap
```
Comment (simPod) [inline src/Utils/ASTDefinitionBuilder.php:62]:
```suggestion
     * @param array<string, Node&TypeDefinitionNode> $typeDefinitionsMap
```

**Review by simPod — APPROVED** [review]
👌🏾

File: src/Utils/SchemaExtender.php (hunk 1 above)
Comment (spawnia) [reply src/Utils/SchemaExtender.php:621]: I don't see how, given the index is not used in the `array_map()`

Comment (simPod) [reply src/Utils/SchemaExtender.php:621]: true, it's not `+`...

File: src/Utils/ASTDefinitionBuilder.php [hunk 4]
```diff
@@ -41,28 +43,28 @@
...
     private $resolveType;
 
-    /** @var Type[] */
+    /** @var array<string, Type> */
     private $cache;
 
     /**
-     * @param Node[] $typeDefinitionsMap
-     * @param bool[] $options
+     * @param array<string, Node&TypeDefinitionNode> $typeDefinitionsMap
```
Comment (spawnia) [inline src/Utils/ASTDefinitionBuilder.php:62]:
```suggestion
     * @param array<string, (Node&TypeDefinitionNode)> $typeDefinitionsMap
```

**Review by vladar — APPROVED** [review]
Nice change 👍 But do you think it is safe to publish as a patch? I am a bit hesitant. Inclining to include it in the next minor.

**Comment by spawnia** [comment]
The actual code changes are very minor and would be best classified as bug fixes. If anything, it would uncover implementations which are already bugged in some way.

**Comment by vladar** [comment]
We have conflicts now after merging #740. Mind fixing? Then we can ship it and safely publish it in the new minor 🚢

**Comment by spawnia** [comment]
@vladar done 🚢

**Review by vladar — APPROVED** [review]
:shipit:

**Comment by shmax** [comment]
@spawnia @vladar  Great work, guys. 👍 By the way, sorry I haven't been very active lately; work has gotten a bit hectic, and my Big Project using graphql is basically done, but please don't hesitate to ping me if you would like me to look at any issues.

**Pull request merged**